### PR TITLE
GVRf Vuforia Integration to support AR features

### PR DESCRIPTION
**Pull request by eesha19**
../GVRf/Sample/AR_SampleApp is a sample app that illustrates AR features in Gvrf with minimum changes in framework. If we avoid doing these changes in the framework, there are significant issues such as: Viewport settings will be incorrect with the tracked object going outside the preview window, the aspect ratio can't also be matched to provide a seamless direct AR type experience. The frame rates in our implementation are also good.

### CODE REVIEW

**Comment by lokeshguddu** [comment]
Hi @gaurav-mcl  @thomasflynn ,

Here is the AR integration that we have implemented. Please have a look and give your feedback.

**Comment by thomasflynn** [comment]
quick feedback.  if you look at the diffs here on github, at least for renderer.cpp (the only file i've looked at so far), it shows that pretty much _every_ _single_ line has changed because the whitespace at the beginning of the line changed for every line.  makes it difficult to see the real diff.

**Comment by thomasflynn** [comment]
ok, this is a pretty large check-in.  will take some time to review.  it still has a VR/AR mode switch that i would prefer not to have in the framework.  i'd prefer that the framework exposed whatever was necessary so that apps could do what they needed.  perhaps expose setting the viewport?  i'll see what i can figure out after i analyze it some more.

in the meantime, please do fix the whitespacing and submit another patch so that the diffs are easier to read.

**Comment by eesha19** [comment]
Hi @thomasflynn ,
I have fixed the white spacing and updated the changes. Please have a look.

File: GVRf/Framework/jni/engine/renderer/renderer.cpp [hunk 1]
```diff
@@ -160,7 +186,15 @@ void Renderer::renderCamera(Scene* scene, Camera* camera, int framebufferId,
         }
 
         glBindFramebuffer(GL_FRAMEBUFFER, framebufferId);
-        glViewport(viewportX, viewportY, viewportWidth, viewportHeight);
+
+        // ---------------------------------------------AR------------------------{
+        if (setIsAr != 0)
+            glViewport(viewportX + ar_offsets[0], viewportY + ar_offsets[1],
```
Comment (chenchao1407) [inline GVRf/Framework/jni/engine/renderer/renderer.cpp]: Please add bracket here and for the else clause even though it is one line call.

File: GVRf/Framework/jni/objects/material_jni.cpp [hunk 2]
```diff
@@ -219,4 +232,41 @@ Material* material = reinterpret_cast<Material*>(jmaterial);
...
+Material* material = reinterpret_cast<Material*>(jmaterial);
+const char* char_key = env->GetStringUTFChars(key, 0);
+std::string native_key = std::string(char_key);
+jfloat* AR_offsets = env->GetFloatArrayElements(arOffsets,0);
+glm::vec4 mat(AR_offsets[0], AR_offsets[1], AR_offsets[2], AR_offsets[3]);
+material->setAROffsets(native_key, mat);
+env->ReleaseStringUTFChars(key, char_key);
+env->ReleaseFloatArrayElements(arOffsets,AR_offsets,0);
+}
+// ---------------------------------------------AR------------------------}
```
Comment (chenchao1407) [inline GVRf/Framework/jni/objects/material_jni.cpp]: ctrl + alt + F here. Thanks.

File: GVRf/Framework/src/org/gearvrf/GVRMaterial.java [hunk 3]
```diff
@@ -70,7 +70,10 @@
     private int mShaderFeatureSet;
     private GVRMaterialShaderId shaderId;
     final private Map<String, GVRTexture> textures = new HashMap<String, GVRTexture>();
+    // ---------------------------------------------AR------------------------{
+    public static float ar_offsets[] = new float[4];
```
Comment (chenchao1407) [inline GVRf/Framework/src/org/gearvrf/GVRMaterial.java]: Naming should follow camel naming in Java. Thanks.

Also I am not considering making it public is a good practice for this parameter.

File: GVRf/Framework/src/org/gearvrf/GVRMaterial.java [hunk 4]
```diff
@@ -214,6 +219,36 @@ public void setMainTexture(Future<GVRTexture> texture) {
...
+        setAROffsets(MAIN_TEXTURE, ar_offsets);
+    }
+
+    public void setAR(String key, int ar) {
+        checkStringNotNullOrEmpty("key", key);
+        NativeMaterial.setAR(getNative(), key, ar);
+    }
+
+    public void setMVPmatrix(String key, float MVPmatrix[]) {
+        checkStringNotNullOrEmpty("key", key);
```
Comment (chenchao1407) [inline GVRf/Framework/src/org/gearvrf/GVRMaterial.java]: The naming for setMVPmat and setMVPmatrix, also with setAR and setATOffsets are really confusing for developers to understand.

File: GVRf/Sample/AR_SampleApp/jni/VR/VRCamera.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+
+	//	memcpy( newViewMat, newViewMat4, 16 * 4);
+
+	/*	matrixMultiply(newViewMat, rotMat, (float*)mViewMat.asArray());
+	matrixInverse(newViewMat);
+
+
+	memcpy((float*)mViewMat.asArray(), newViewMat, sizeof(float) * 16);*/
+}
```
Comment (chenchao1407) [inline GVRf/Sample/AR_SampleApp/jni/VR/VRCamera.cpp]: Too many commented codes here....

File: GVRf/Sample/AR_SampleApp/jni/VR/VRCamera.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+		float mUp[3];
+
+		GVR::mat4 mViewMat;
+		GVR::mat4 mProjMat;
+
+		float mFov;
+		float mAspect;
+		float mNear;
+		float mFar;
+
```
Comment (chenchao1407) [inline GVRf/Sample/AR_SampleApp/jni/VR/VRCamera.h]: Hi. Please take a look at http://gearvrf.org/bin/view/GearVRF/GearVRfWikiGetInvolved#CodeGuide

This is the coding guideline for GVRF patches. Thanks

File: GVRf/Sample/AR_SampleApp/jni/VR/VRCurvedMesh.cpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+		float x = (it->y + 1.0f) * 0.5f;
+		float y = 1.0f - (it->x +1.0f) * 0.5f;
+
+	//	x += (0.1f * sin( 3.14f * (y) ) * it->y);
+
+		android::vec2 val = android::vec2(x, y) ;
+		uvs.push_back(val);
+	}*/
+
```
Comment (chenchao1407) [inline GVRf/Sample/AR_SampleApp/jni/VR/VRCurvedMesh.cpp]: Please clean this commented codes. Thanks

File: GVRf/Sample/AR_SampleApp/jni/VR/VRSceneManager.cpp [hunk 8]
```diff
@@ -0,0 +1,860 @@
...
+			}
+			if(gloHeader->mMaxAnimFrame > 50000) {
+				VR_LOGE("gloHeader->mMaxAnimFrame > 50000 so clamping it to 50000");
+				gloHeader->mMaxAnimFrame = 50000;
+			}
+			if(gloHeader->fileSize == 0) {
+				VR_LOGE("gloHeader->fileSize = 0 so clamping it to 3000");
+				return false;
+			}*/
+		}
```
Comment (chenchao1407) [inline GVRf/Sample/AR_SampleApp/jni/VR/VRSceneManager.cpp]: Same here...

File: GVRf/Sample/AR_SampleApp/jni/VR/VRSceneManager.cpp [hunk 9]
```diff
@@ -0,0 +1,860 @@
...
+        {
+            int geomIndx = efxNode->mJoinGeomIndx[p];
+            int boneIndx = efxNode->mJoinGeomIndx[p + 1];
+           /* float *tmp = scn->mBoneAnimMat[geomIndx];           
+            tmp = tmp + 16 * boneIndx;
+            matMul(tmpMat, boneMat, globMat);
+            matMul(tmp, bindShapeMat, tmpMat);
+            boneMat += 16;
+            bindShapeMat += 16;*/
+	     tmpMat = boneMat * globMat;
```
Comment (chenchao1407) [inline GVRf/Sample/AR_SampleApp/jni/VR/VRSceneManager.cpp]: Same here..

File: GVRf/Sample/AR_SampleApp/jni/VR/VRSceneManager.cpp [hunk 10]
```diff
@@ -0,0 +1,860 @@
...
+		isSceneLoaded = mChairs->loadFromFile("/data/log/vr/chair.bin"); // chair.bin //SpaceShip.bin
+	#endif 
+		if(isSceneLoaded) {
+			this->addToRenderer(mChairs);
+		} else {
+
+		VR_LOGE("VR Error :: VRSceneRenderer::int() , Couldn't load the chair [%x] scene !!\n", mChairs);
+
+		}*/
+
```
Comment (chenchao1407) [inline GVRf/Sample/AR_SampleApp/jni/VR/VRSceneManager.cpp]: Here...

File: GVRf/Sample/AR_SampleApp/jni/VR/VRSceneManager.h [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+#include <vector>
+#include <GLES2/gl2.h>
+#include <GLES2/gl2ext.h>
+#include "kSensor/math/matrix.hpp"
+#include "VRCamera.h"
+#include "VRShaderUtil.h"
+
+namespace GVR {
+
+	enum VRSceneType {
```
Comment (chenchao1407) [inline GVRf/Sample/AR_SampleApp/jni/VR/VRSceneManager.h]: Not quite sure but the formatting of this file looks quite weird to me. Thanks.

File: GVRf/Sample/AR_SampleApp/CopyVuforiaFiles.xml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project name="VuforiaSamples" default="CopyVuforiaSo">
+    <target name="CopyVuforiaSo">
+        <copy overwrite="true" todir="C:\adt\adt-bundle-windows-x86-20131030\vuforia-sdk-android-3-0-9\build\lib\armeabi-v7a">
+			<fileset dir="C:\adt\adt-bundle-windows-x86-20131030\vuforia-sdk-android-3-0-9\build\lib\armeabi-v7a"/>
```
Comment (chenchao1407) [inline GVRf/Sample/AR_SampleApp/CopyVuforiaFiles.xml]: The configuration here make the sample unable to build. I believe the todir parameter and dir parameter are using your local url...

**Comment by chenchao1407** [comment]
Just took a quick review. I mean, too many naming, formatting issues and commented codes there. I tried to make comment for every related issue but found it is a little bit frustrating to make comments for almost most of files. Please take a look at : http://gearvrf.org/bin/view/GearVRF/GearVRfWikiGetInvolved#CodeGuide to try to follow the coding guidelines for your codes. Thanks in advance.

Also, I just feel a little weird about the design of this app.  I am not sure if it will look perfect to port another render engine with its own distorter, renderer here as just a sample. Anyway, tried to run it but a configuration file issue found. Please correct it and also with formatting issues, commented codes issues. Thanks in advance.

**Comment by thomasflynn** [comment]
Yeah, the sample app can't even be imported.   Something about C:/adt/adt-bundle-windows-x86-20131030/vuforia-sdk-android-3-0-9/build/lib/armeabi-v7a does not exist.  Please fix.

**Comment by gaurav-mcl** [comment]
1. Please remove CopyVuforiaFiles.xml from your app because it creates a dependency on vuforia sdk folder that needs to be installed. @thomasflynn and @chenchao1407 are having issues because they don't have vuforia sdk installed. Better to package libVuforia.so directly in your app so that users don't need to install the vuforia sdk.
For this, please copy libVuforia.so from vuforia-sdk-android-3-0-9/build/lib/armeabi-v7a to your app's jni folder and edit the Android.mk file to add following lines:

```
# adding prebuilt vuforia library
include $(CLEAR_VARS)

LOCAL_MODULE    := libVuforia
LOCAL_SRC_FILES := libVuforia.so

include $(PREBUILT_SHARED_LIBRARY)
```

2. Similarly please remove the classpath entry corresponding to Vuforia.jar because that again has dependency on external vuforia sdk folder. Instead please copy vuforia-sdk-android-3-0-9/build/java/vuforia/Vuforia.jar to your app's lib folder.

**Comment by gaurav-mcl** [comment]
Also please remove the following binary files from the commit because they get created during the build process :

* Framework/libs/armeabi-v7a/libassimp.so
* Framework/libs/armeabi-v7a/libgvrf.so
* Framework/libs/armeabi-v7a/libvrapi.so

**Comment by thomasflynn** [comment]
you can look at the gvr-vuforia sample for how to include the vuforia files with your app.

**Comment by thomasflynn** [comment]
it looks like maybe there's a number files you maybe didn't intend to check-in?  I see a number of textures that aren't referenced and a lot of cpp code that doesn't seem to be loaded in the associated java files.

**Comment by eesha19** [comment]
Hi @chenchao1407 , @thomasflynn , @gaurav-mcl 
I have updated the project with the required changes in the framework and application. Also, you will be able to import and run it successfully (dependencies removed). 
Please have a look and see if this suits your requirements.
Thanks

**Comment by thomasflynn** [comment]
thanks for the updates.  it looks as though you still have some binaries checked in that shouldn't be.  for example:  libgvrf.so, libassimp.so, libvrapi.so.

**Comment by liaxim** [comment]
Since this is in indeterminate state, planning to close it after a day or two. You are welcome to open a new one once the change is brought up to date.

**Pull request closed without merging**